.github/workflows/container_image.yaml

```diff
@@ -19,14 +19,6 @@ jobs:
     - run: |
         docker login -u="${{ secrets.QUAY_USERNAME }}" -p="${{ secrets.QUAY_TOKEN }}" quay.io
 
-        # Build and push "image:commit_sha"
-        docker build -t "${{ env.image }}:${GITHUB_SHA}" \
-          && docker push "${{ env.image }}:${GITHUB_SHA}"
-
-        # Tag and push "image:branch"
-        docker tag "${{ env.image }}:${GITHUB_SHA}" "${{ env.image }}:${GITHUB_REF_NAME}" \
-          && docker push "${{ env.image }}:${GITHUB_REF_NAME}"
-
-        # Tag and push "image:latest"
-        docker tag "${{ env.image }}:${GITHUB_SHA}" "${{ env.image }}" \
-          && docker push "${{ env.image }}"
+        docker build -t ${{ env.image }}:main -t ${{ env.image }}:latest .
+        docker build --label quay.expires-after=4w -t ${{ env.image }}:${GITHUB_SHA::7} .
+        docker push --all-tags ${{ env.image }}
```

Comment:
TIL `--all-tags`!

I'm not 100% sure, but I don't think this is going to do what you expect. My expectation is that this is going to push all tags of a *specific* image, but we've built 2 images here.

My expectation (but again not 100% sure) is that an untagged image may implicitly refer to `latest`, in which case this would only push the first image, i.e. the `main` and `latest` tags[^1].

My recommendation would be:
* Specify the tag explicitly
* Push both images
* Consider dropping the `main` tag, as nobody's likely to use it over the more commonly used `latest`.

This would also mean each image ends up with only a single tag, so `--all-tags` would no longer be required. i.e.

```
        docker build -t ${{ env.image }}:latest .
        docker push ${{ env.image }}:latest
        docker build --label quay.expires-after=4w -t ${{ env.image }}:${GITHUB_SHA::7} .
        docker push ${{ env.image }}:${GITHUB_SHA::7}
```

If you wanted to keep the `main` tag the first build you'd add the tag back to the first build and add `--all-tags` to the first push.

[^1]: Do we need both of those?

Reply:
>this is going to push all tags of a specific image

Right. Looks like there is a 100% chance you're right. Let me fix that.

Reply:
You don't want `latest` to start bouncing between branches, because that makes it useless. I would expect `latest` to **always** mean the build of the most recent commit on main. When we have release branches, I would expect to have a different *latest* tag for each branch, e.g. `release-0.1-latest`.